src/shacl/ShaclParser.ts

```diff
@@ -0,0 +1,583 @@
+      {
+        ALT: () => this.SUBRULE(this.shaclOneOrMorePath),
+      },
+      {
+        ALT: () => this.SUBRULE(this.shaclZeroOrOnePath),
+      },
+    ]);
+  });
+
+  shaclPredicatePath = this.RULE('shaclPredicatePath', () => {
+    this.OR([
+      {
+        ALT: () => this.SUBRULE(this.iri),
+      },
+      {
+        // This case does not seem to be allowed by the SHACL spec, but the
+        // online W3C validator accepts one level of parens wrapping the IRI.
+        ALT: () => {
+          this.CONSUME(turtleTokenMap.LParen);
+          this.SUBRULE1(this.iri);
+          this.CONSUME(turtleTokenMap.RParen);
+        },
+      },
+    ]);
+  });
+
+  shaclSequencePath = this.RULE('shaclSequencePath', () => {
+    this.CONSUME(turtleTokenMap.LParen);
+    this.SUBRULE(this.shaclPropertyPathPath);
+    this.AT_LEAST_ONE(() => this.SUBRULE1(this.shaclPropertyPathPath));
+    this.OPTION(() => this.CONSUME(turtleTokenMap.Semicolon));
+    this.CONSUME(turtleTokenMap.RParen);
+  });
+
+  shaclAlternativePath = this.RULE('shaclAlternativePath', () => {
+    this.CONSUME(turtleTokenMap.LBracket);
+    this.CONSUME(this.shaclTokenMap.SHACL_alternativePath);
+    this.SUBRULE(this.shaclPropertyPathPath); // This does not match the SHACL spec, but it does match the test cases, which violate the spec. ;_;
+    this.OPTION(() => this.CONSUME(turtleTokenMap.Semicolon));
+    this.CONSUME(turtleTokenMap.RBracket);
+  });
+
+  shaclInversePath = this.RULE('shaclInversePath', () => {
+    this.CONSUME(turtleTokenMap.LBracket);
+    this.CONSUME(this.shaclTokenMap.SHACL_inversePath);
+    this.SUBRULE(this.shaclPropertyPathPath);
+    this.OPTION(() => this.CONSUME(turtleTokenMap.Semicolon));
+    this.CONSUME(turtleTokenMap.RBracket);
+  });
+
+  shaclZeroOrMorePath = this.RULE('shaclZeroOrMorePath', () => {
+    this.CONSUME(turtleTokenMap.LBracket);
+    this.CONSUME(this.shaclTokenMap.SHACL_zeroOrMorePath);
+    this.SUBRULE(this.shaclPropertyPathPath);
+    this.OPTION(() => this.CONSUME(turtleTokenMap.Semicolon));
+    this.CONSUME(turtleTokenMap.RBracket);
+  });
+
+  shaclOneOrMorePath = this.RULE('shaclOneOrMorePath', () => {
+    this.CONSUME(turtleTokenMap.LBracket);
+    this.CONSUME(this.shaclTokenMap.SHACL_oneOrMorePath);
+    this.SUBRULE(this.shaclPropertyPathPath);
+    this.OPTION(() => this.CONSUME(turtleTokenMap.Semicolon));
+    this.CONSUME(turtleTokenMap.RBracket);
+  });
+
+  shaclZeroOrOnePath = this.RULE('shaclZeroOrOnePath', () => {
+    this.CONSUME(turtleTokenMap.LBracket);
+    this.CONSUME(this.shaclTokenMap.SHACL_zeroOrOnePath);
+    this.SUBRULE(this.shaclPropertyPathPath);
+    this.OPTION(() => this.CONSUME(turtleTokenMap.Semicolon));
+    this.CONSUME(turtleTokenMap.RBracket);
+  });
+
+  shaclLiteralConstraint = this.RULE('shaclLiteralConstraint', () => {
+    this.OR([
+      {
+        ALT: () => this.SUBRULE(this.shaclIntConstraint),
+      },
+      // TODO: Some specificity here is possibly unnecessary.
```

Comment:
what do you mean?

Reply:
I'll try to remember and update the comment. 😛